src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/LaunchSettingsProviderTests.cs

```diff
@@ -0,0 +1,489 @@
+        {
+
+            IFileSystemMoq moqFS = new IFileSystemMoq();
+
+            var provider = GetLaunchSettingsProvider(moqFS);
+
+            // No settings  file, should add the default profile
+            await provider.UpdateProfilesAsyncTest(null);
+            Assert.Equal(1, provider.CurrentSnapshot.Profiles.Count);
+            Assert.Equal("Project", provider.CurrentSnapshot.ActiveProfile.CommandName);
+
+            // Add one
+            moqFS.WriteAllText(provider.LaunchSettingsFile, JsonString1);
+
+            await provider.UpdateProfilesAsyncTest(null);
+            Assert.Equal(4, provider.CurrentSnapshot.Profiles.Count);
+            Assert.Equal(0, provider.CurrentSnapshot.GlobalSettings.Count);
+            Assert.Equal("IIS Express", provider.CurrentSnapshot.ActiveProfile.Name);
+            
+            // don't change file on disk, just active one
+            await provider.UpdateProfilesAsyncTest("Docker");
+            Assert.Equal(4, provider.CurrentSnapshot.Profiles.Count);
+            Assert.Equal(0, provider.CurrentSnapshot.GlobalSettings.Count);
+            Assert.Equal("Docker", provider.CurrentSnapshot.ActiveProfile.Name);
+
+            // Update with bad json, profiles should remain stable
+            moqFS.WriteAllText(provider.LaunchSettingsFile, BadJsonString);
+            await provider.UpdateProfilesAsyncTest("Docker");
+            Assert.Equal(4, provider.CurrentSnapshot.Profiles.Count);
+            Assert.Equal(0, provider.CurrentSnapshot.GlobalSettings.Count);
+            Assert.Equal("Docker", provider.CurrentSnapshot.ActiveProfile.Name);
+            
+            // Clear snapshot, this time an single error profile should be added
+            provider.SetCurrentSnapshot(null);
+            await provider.UpdateProfilesAsyncTest("Docker");
+            Assert.Equal(1, provider.CurrentSnapshot.Profiles.Count);
+            Assert.Equal(LaunchSettingsProvider.ErrorProfileCommandName, provider.CurrentSnapshot.ActiveProfile.CommandName);
+
+        }
+
+        [Fact]
+        public void LaunchSettingsProvider_SettingsFileHasChangedTests()
+        {
+            IFileSystemMoq moqFS = new IFileSystemMoq();
+            var provider = GetLaunchSettingsProvider(moqFS);
+
+            // No settings  file
+            Assert.True(provider.SettingsFileHasChangedTest());
+            moqFS.WriteAllText(provider.LaunchSettingsFile, JsonString1);
+
+            Assert.True(provider.SettingsFileHasChangedTest());
+            provider.LastSettingsFileSyncTimeTest = moqFS.LastFileWriteTime(provider.LaunchSettingsFile);
+            Assert.False(provider.SettingsFileHasChangedTest());
+        }
+
+
+        [Fact]
+        public void LaunchSettingsProvider_ReadProfilesFromDiskTests()
+        {
+
+            IFileSystemMoq moqFS = new IFileSystemMoq();
+            var provider = GetLaunchSettingsProvider(moqFS);
+
+            // Test without an existing file. Should throw
```

Comment:
Can we break up these tests please? Lots of different failure modes in a single test leads to hard test failures to diagnose.